src/data/int/log.lean

```diff
@@ -0,0 +1,294 @@
+  cases le_total 1 r with hr1 hr1,
+  { rw log_of_one_le_right _ hr1,
+    refine le_trans _ (nat.floor_le hr.le),
+    rw [zpow_coe_nat, ←nat.cast_pow, nat.cast_le],
+    exact nat.pow_log_le_self hb (nat.floor_pos.mpr hr1) },
+  { rw [log_of_right_le_one _ hr1, zpow_neg, zpow_coe_nat, ← nat.cast_pow],
+    apply inv_le_of_inv_le hr,
+    refine (nat.le_ceil _).trans (nat.cast_le.2 _),
+    exact nat.le_pow_clog hb _ },
+end
+
+lemma lt_zpow_succ_log_self {b : ℕ} {r : R} (hb : 1 < b) (hr : 0 < r) :
+  r < (b : R) ^ (log b r + 1) :=
+begin
+  cases le_or_lt 1 r with hr1 hr1,
+  { rw log_of_one_le_right _ hr1,
+    rw [int.coe_nat_add_one_out, zpow_coe_nat, ←nat.cast_pow],
+    apply nat.lt_of_floor_lt,
+    exact nat.lt_pow_succ_log_self hb _, },
+  { rw log_of_right_le_one _ hr1.le,
+    have hcri : 1 < r⁻¹ := one_lt_inv hr hr1,
+    have : 1 ≤ nat.clog b ⌈r⁻¹⌉₊ :=
+      nat.succ_le_of_lt (nat.clog_pos hb $ nat.one_lt_cast.1 $ hcri.trans_le (nat.le_ceil _)),
+    rw [neg_add_eq_sub, ←neg_sub, ←int.coe_nat_one, ← int.coe_nat_sub this,
+      zpow_neg, zpow_coe_nat, lt_inv hr (pow_pos (nat.cast_pos.mpr $ zero_lt_one.trans hb) _),
+      ←nat.cast_pow],
+    refine nat.lt_ceil.1 _,
+    exact (nat.pow_pred_clog_lt_self hb $ nat.one_lt_cast.1 $ hcri.trans_le $ nat.le_ceil _), }
+end
+
+@[simp] lemma log_zero_right (b : ℕ) : log b (0 : R) = 0 :=
+log_of_right_le_zero b le_rfl
+
+@[simp] lemma log_one_right (b : ℕ) : log b (1 : R) = 0 :=
+by rw [log_of_one_le_right _ le_rfl, nat.floor_one, nat.log_one_right, int.coe_nat_zero]
+
+lemma log_zpow {b : ℕ} (hb : 1 < b) (z : ℤ) : log b (b ^ z : R) = z :=
+begin
+  obtain ⟨n, rfl | rfl⟩ := z.eq_coe_or_neg,
+  { rw [log_of_one_le_right _ (one_le_zpow_of_nonneg _ $ int.coe_nat_nonneg _),
+      zpow_coe_nat, ←nat.cast_pow, nat.floor_coe, nat.log_pow hb],
+    exact_mod_cast hb.le, },
+  { rw [log_of_right_le_one _ (zpow_le_one_of_nonpos _ $ neg_nonpos.mpr (int.coe_nat_nonneg _)),
+      zpow_neg, inv_inv, zpow_coe_nat, ←nat.cast_pow, nat.ceil_coe, nat.clog_pow _ _ hb],
+    exact_mod_cast hb.le, },
+end
+
+@[mono] lemma log_mono_right {b : ℕ} {r₁ r₂ : R} (h₀ : 0 < r₁) (h : r₁ ≤ r₂) :
+  log b r₁ ≤ log b r₂ :=
+begin
+  cases le_or_lt b 1 with hb hb,
+  { rw [log_of_left_le_one hb, log_of_left_le_one hb], },
+  cases le_total r₁ 1 with h₁ h₁; cases le_total r₂ 1 with h₂ h₂,
+  { rw [log_of_right_le_one _ h₁, log_of_right_le_one _ h₂, neg_le_neg_iff, int.coe_nat_le],
+    exact nat.clog_mono_right _ (nat.ceil_mono $ inv_le_inv_of_le h₀ h), },
+  { rw [log_of_right_le_one _ h₁, log_of_one_le_right _ h₂],
+    exact (neg_nonpos.mpr (int.coe_nat_nonneg _)).trans (int.coe_nat_nonneg _) },
+  { obtain rfl := le_antisymm h (h₂.trans h₁), refl, },
+  { rw [log_of_one_le_right _ h₁, log_of_one_le_right _ h₂, int.coe_nat_le],
+    exact nat.log_mono_right (nat.floor_mono h), },
+end
+
+variables (R)
+
+/-- Over suitable subtypes, `zpow` and `int.log` form a galois coinsertion -/
+def zpow_log_gci {b : ℕ} (hb : 1 < b) :
```

Comment:
The (confusing?) naming convention is `gc` for `galois_connection` and `gi` for `galois_insertion`/`galois_coinsertion`, because at most one of them will be true before you have something more interesting to say.
```suggestion
def zpow_log_gi {b : ℕ} (hb : 1 < b) :
```